### PR TITLE
UCT/API: Add memory type extensions to uct_md_mem_alloc /uct_mem_alloc

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What ?
UCT API to allocate memory on a specified memory type. 
## Why ?
Current implementation of uct_mem_alloc or uct_md_mem_alloc is used to implement memory pool allocation in UCP. These routines don't provide an option to allocate memory on GPU device memory which is needed to implement efficient pipelined transfers for the purposes of 1. user managed memory buffers 2. user pinned memory buffers when devices have BAR size < physical limit. 

## How ?
Adds variants of `uct_md_mem_alloc` and `uct_mem_alloc` to accept `mem_type` parameter which indicates the type of memory that needs to be allocated as opposed to defaulting to allocation of host memory.

### CODE REVIEW

**Comment by Akshay-Venkatesh** [comment]
cc @ssaulters @souravzzz

**Comment by brminich** [comment]
ok to test

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -2198,6 +2230,56 @@ ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
...
+ * @param [in]     num_methods Length of 'methods' array.
+ * @param [in]     mds         Array of memory domains to attempt to allocate
+ *                             the memory with, for MD allocation method. If
+ *                             none of the provided mds can allocate memory of
+ *                             the specified memory type, then allocation fails
+ *                             if MD allocation method is requested.
+ * @param [in]     num_mds     Length of 'mds' array. May be empty, in such case
+ *                             'mds' may be NULL, and MD allocation method will
+ *                             be skipped.
+ * @param [in]     mem_type    Type of memory to be allocated.
```
Comment (brminich) [inline src/uct/api/uct.h:2269]: does it make `methods` parameter irrelevant?
Also maybe define extentable function, using parameters structure - `uct_mem_alloc_param_t`?

Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:2269]: So far methods would make sense for the case when mem_type = TYPE_HOST. But there are also a subset of such methods available in cuda that could used for future purposes.

Defining an extensible function makes sense. Will iterate on this. Thanks!

**Comment by yosefe** [comment]
bot:pipe:retest

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -2082,6 +2083,37 @@ ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
...
+ *                             gaurantee an allocation starting at the specified
+ *                             address, an appropritate error is returned.
+ * @param [in]     mem_type    Type of memory to be allocated.
+ *                             E.g. UCS_MEMORY_TYPE_HOST, UCS_MEMORY_TYPE_CUDA,
+ *                             UCS_MEMORY_TYPE_ROCM, etc.
+ * @param [in]     name        Name of the allocated region, used to track memory
+ *                             usage for debugging and profiling.
+ * @param [out]    memh_p      Filled with handle for allocated region.
+ */
+ucs_status_t uct_md_mem_alloc_mem_type(uct_md_h md, size_t *length_p,
```
Comment (yosefe) [inline src/uct/api/uct.h:2111]: instead of adding new API, we should extend uct_md_mem_alloc() with params(and field_mask) struct, which will also take the memory type

Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h:2111]: @yosefe @brminich Adding a draft of extending uct_md_mem_alloc/uct_mem_alloc borrowing ucp_request_param_t additions in UCP. LMK if this is along the lines we should change
```diff
+/**
+ * @ingroup UCT_MD
+ * @brief UCT allocation parameter specification field
+ *
+ * The enumeration allows specifying which fields in @ref uct_mem_alloc_param_t
+ * are present.
+ */
+typedef enum {
+    UCT_ALLOC_ATTR_FIELD_FLAGS    = UCS_BIT(0) /**< Allocation flags field*/
+    UCT_ALLOC_ATTR_FIELD_MEM_TYPE = UCS_BIT(1) /**< Memory type field */
+} uct_alloc_attr_t;
+
+
+/**
+ * @ingroup UCT_MD
+ * @brief Allocation characteristics parameter passed to @ref uct_mem_alloc,
+ *        @ref uct_md_mem_alloc
+ *
+ * The structure @ref uct_mem_alloc_param_t is used to specify the memory type
+ * of the allocation, and flags indicating other characteristics of the
+ * allocation.
+ */
+typedef struct {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_alloc_attr_t. Fields not specified in this mask will be
+     * ignored
+     */
+    uint32_t       alloc_attr_mask;
+
+    /**
+     * Type of memory to be allocated.
+     * E.g. UCS_MEMORY_TYPE_HOST, UCS_MEMORY_TYPE_CUDA, UCS_MEMORY_TYPE_ROCM.
+     */
+    ucs_memory_type_t mem_type;
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned flags;
+
+} uct_mem_alloc_param_t;
+
+
 /**
  * @ingroup UCT_MD
  * @brief Allocate memory for zero-copy sends and remote access.
@@ -2073,13 +2118,15 @@ ucs_status_t uct_md_query(uct_md_h md, uct_md_attr_t *md_attr);
  *                             return, filled with the actual size that was allocated,
  *                             which may be larger than the one requested. Must be >0.
  * @param [in,out] address_p   The address
- * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
+ * @param [in]     param       Memory allocation characteristics, see
+ *                             @ref uct_mem_alloc_param_t.
  * @param [in]     name        Name of the allocated region, used to track memory
  *                             usage for debugging and profiling.
  * @param [out]    memh_p      Filled with handle for allocated region.
  */
 ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *name, uct_mem_h *memh_p);
+                              uct_mem_alloc_param_t *param, const char *name,
+                              uct_mem_h *memh_p);
 
 
 /**
@@ -2180,7 +2227,8 @@ ucs_status_t uct_md_detect_memory_type(uct_md_h md, const void *addr,
  *                             cannot be made, the allocation request fails.
  * @param [in]     min_length  Minimal size to allocate. The actual size may be
  *                             larger, for example because of alignment restrictions.
- * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
+ * @param [in]     param       Memory allocation characteristics, see
+ *                             @ref uct_mem_alloc_param_t.
  * @param [in]     methods     Array of memory allocation methods to attempt.
  * @param [in]     num_methods Length of 'methods' array.
  * @param [in]     mds         Array of memory domains to attempt to allocate
@@ -2192,7 +2240,8 @@ ucs_status_t uct_md_detect_memory_type(uct_md_h md, const void *addr,
  * @param [out]    mem         In case of success, filled with information about
  *                              the allocated memory. @ref uct_allocated_memory_t.
  */
-ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
+ucs_status_t uct_mem_alloc(void *addr, size_t min_length,
+                           uct_mem_alloc_param_t *param,
                            uct_alloc_method_t *methods, unsigned num_methods,
                            uct_md_h *mds, unsigned num_mds, const char *name,
                            uct_allocated_memory_t *mem);

```

Comment (brminich) [reply src/uct/api/uct.h]: looks good to me. Maybe also move `mds` and `num_mds` of `uct_mem_alloc` to params, because they are optional?

File: src/uct/api/uct.h [hunk 3]
```diff
@@ -2061,6 +2062,49 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_alloc_attr_t. Fields not specified in this mask will be
+     * ignored
+     */
+    uint32_t       alloc_attr_mask;
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned flags;
```
Comment (yosefe) [inline src/uct/api/uct.h:2098]: 1. align on column
2. 2. pls specify which field_mask flag is enabling it

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -2061,6 +2062,49 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+ * of the allocation, and flags indicating other characteristics of the
+ * allocation.
+ */
+typedef struct {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_alloc_attr_t. Fields not specified in this mask will be
+     * ignored
+     */
+    uint32_t       alloc_attr_mask;
```
Comment (yosefe) [inline src/uct/api/uct.h:2093]: uint64_t field_mask

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -2061,6 +2062,49 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned flags;
+
+    /**
+     * Type of memory to be allocated.
+     * E.g. UCS_MEMORY_TYPE_HOST, UCS_MEMORY_TYPE_CUDA, UCS_MEMORY_TYPE_ROCM.
```
Comment (yosefe) [inline src/uct/api/uct.h:2102]: 1. e.g is redundant since ucs_memory_type_t  is defined
2. pls specify which field_mask flag is enabling it

File: src/uct/api/uct.h [hunk 6]
```diff
@@ -2192,7 +2239,8 @@ ucs_status_t uct_md_detect_memory_type(uct_md_h md, const void *addr,
  * @param [out]    mem         In case of success, filled with information about
  *                              the allocated memory. @ref uct_allocated_memory_t.
  */
-ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
+ucs_status_t uct_mem_alloc(void *addr, size_t min_length,
+                           uct_mem_alloc_param_t *param,
```
Comment (yosefe) [inline src/uct/api/uct.h:2243]: pls move ALL function parameters inside uct_mem_alloc_param_t

File: src/uct/api/uct.h [hunk 7]
```diff
@@ -2061,6 +2062,49 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
 
+/**
+ * @ingroup UCT_MD
+ * @brief UCT allocation parameter specification field
+ *
+ * The enumeration allows specifying which fields in @ref uct_mem_alloc_param_t
+ * are present.
+ */
+typedef enum {
+    UCT_ALLOC_ATTR_FIELD_FLAGS    = UCS_BIT(0), /**< Allocation flags field*/
```
[the remark this answers is not in the document]
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: done

File: src/uct/api/uct.h [hunk 8]
```diff
@@ -2061,6 +2062,49 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+    UCT_ALLOC_ATTR_FIELD_MEM_TYPE = UCS_BIT(1)  /**< Memory type field */
+} uct_alloc_attr_t;
```
[the remark this answers is not in the document]
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: done

File: src/uct/api/uct.h (hunk 3 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: Done

File: src/uct/api/uct.h (hunk 4 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: Done

File: src/uct/api/uct.h (hunk 5 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: Will this do? 
https://github.com/openucx/ucx/pull/5204/files#diff-a1f319ca66018bb8d03274d3b1757ec7R2094

File: src/uct/api/uct.h (hunk 6 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: done

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by Akshay-Venkatesh** [comment]
@yosefe Shall we resume working this PR again?

File: src/uct/api/uct.h [hunk 9]
```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+    uint64_t           field_mask;
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned           flags;
+
+    /**
+     * The address. Valid to use with @ref uct_md_mem_alloc
+     */
```
Comment (yosefe) [inline src/uct/api/uct.h:2118]: what is this related to?

File: src/uct/api/uct.h [hunk 10]
```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+
+    /**
+     * Array of memory allocation methods to attempt. Every one of the provided
+     * allocation methods will be used, in turn, to perform the allocation,
+     * until one succeeds. Whenever the MD method is encountered, every one
+     * of the provided MDs will be used, in turn, to allocate the memory, until
+     * one succeeds, or they are exhausted. In this case the next allocation
+     * method from the initial list will be attempted.
+     */
+    uct_alloc_method_t *methods;
```
Comment (yosefe) [inline src/uct/api/uct.h:2147]: const  uct_alloc_method_t *methods;

File: src/uct/api/uct.h [hunk 11]
```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+    /**
+     * Length of 'methods' array
+     */
+    unsigned           num_methods;
+
+    /**
+     * Array of memory domains to attempt to allocate
+     * the memory with, for MD allocation method.
+     */
+    uct_md_h           *mds;
```
Comment (yosefe) [inline src/uct/api/uct.h:2158]: const uct_md_h *mds;

File: src/uct/api/uct.h [hunk 12]
```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+     * of the provided MDs will be used, in turn, to allocate the memory, until
+     * one succeeds, or they are exhausted. In this case the next allocation
+     * method from the initial list will be attempted.
+     */
+    uct_alloc_method_t *methods;
+
+    /**
+     * Length of 'methods' array
+     */
+    unsigned           num_methods;
```
Comment (yosefe) [inline src/uct/api/uct.h:2152]: should put methods and num methods in sub-struct, with one bit in field_mask for both
```
struct {
    const uct_alloc_method_t   *methods;
    unsigned                   count;
} methods;
```

same thing for "mds"

File: src/uct/api/uct.h [hunk 13]
```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+     *  Length of 'mds' array. May be empty, in such case
+     *  'mds' may be NULL, and MD allocation method will
+     *  be skipped.
+     */
+    unsigned           num_mds;
+
+    /**
+     * Type of memory to be allocated.
+     */
+    ucs_memory_type_t  mem_type;
```
Comment (yosefe) [inline src/uct/api/uct.h:2170]: let's order fields this way:
1. address_p
2. length_p
3. mem_type
4. methods
5. mds
6. name

File: src/uct/base/uct_mem.c [hunk 14]
```diff
@@ -77,30 +78,31 @@ ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
     ssize_t huge_page_size;
 #endif
 
-    if (min_length == 0) {
+    if (*param->length_p == 0) {
```
Comment (yosefe) [inline src/uct/base/uct_mem.c:81]: all accesses to param must be after checking field_mask (in multiple locations in this PR; pls fix all of them)

File: src/uct/api/uct.h [hunk 15]
```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+
+    /**
+     * Name of the allocated region, used to track memory
+     * usage for debugging and profiling.
+     */
+    const char         *name;
+} uct_mem_alloc_param_t;
```
Comment (yosefe) [inline src/uct/api/uct.h:2177]: uct_mem_alloc_param**s**_t

File: src/uct/api/uct.h [hunk 16]
```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
...
+
+    /** Enables @ref uct_mem_alloc_param_t::num_mds */
+    UCT_MEM_ALLOC_PARAM_FIELD_NUM_MDS        = UCS_BIT(6),
+
+    /** Enables @ref uct_mem_alloc_param_t::mem_type */
+    UCT_MEM_ALLOC_PARAM_FIELD_MEM_TYPE       = UCS_BIT(7),
+
+    /** Enables @ref uct_mem_alloc_param_t::name */
+    UCT_MEM_ALLOC_PARAM_FIELD_NAME           = UCS_BIT(8)
+} uct_mem_alloc_param_field_t;
```
Comment (yosefe) [inline src/uct/api/uct.h:2095]: uct_mem_alloc_param**s**_field_t

File: src/uct/api/uct.h [hunk 17]
```diff
@@ -2065,17 +2185,12 @@ ucs_status_t uct_md_query(uct_md_h md, uct_md_attr_t *md_attr);
...
- * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
- * @param [in]     name        Name of the allocated region, used to track memory
- *                             usage for debugging and profiling.
+ * @param [in]     param       Memory allocation characteristics, see
+ *                             @ref uct_mem_alloc_param_t.
  * @param [out]    memh_p      Filled with handle for allocated region.
  */
-ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *name, uct_mem_h *memh_p);
+ucs_status_t uct_md_mem_alloc(uct_md_h md, uct_mem_alloc_param_t *param,
```
Comment (yosefe) [inline src/uct/api/uct.h:2192]: const uct_mem_alloc_param_t *param

File: src/uct/base/uct_mem.c [hunk 18]
```diff
@@ -124,18 +126,17 @@ ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
                  * fall-back, because this MD already exposed support for memory
                  * allocation.
                  */
-                alloc_length = min_length;
-                address = addr;
-                status = uct_md_mem_alloc(md, &alloc_length, &address, flags,
-                                          alloc_name, &memh);
+                alloc_length     = *param->length_p;
+                status = uct_md_mem_alloc(md, param, &memh);
```
Comment (yosefe) [inline src/uct/base/uct_mem.c:130]: align = on column

File: src/uct/base/uct_mem.c (hunk 14 above)
Comment (Akshay-Venkatesh) [reply src/uct/base/uct_mem.c:81]: Will address in the next commit.

File: src/uct/api/uct.h (hunk 17 above)
Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: @yosefe Just realized that we can't use const qualifier because param->address_p is allowed to be changed after initially used a hint. Following two examples implement md_mem_alloc and return potentially new address_p values. For this reason we may have to change the param type to [inout] and remove const qualifier.

https://github.com/openucx/ucx/blob/master/src/uct/sm/mm/sysv/mm_sysv.c#L106
https://github.com/openucx/ucx/blob/master/src/uct/sm/mm/posix/mm_posix.c#L509

Comment (yosefe) [reply src/uct/api/uct.h]: not really, address_p is `void**`, not `void*`, we should be able to use const

Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: Oh right. I agree now that const can be used. 

What about setting param type from [in] to [inout] ? Don't we have to make this change given that param->address_p content may be updated by uct_md_mem_alloc?

Comment (yosefe) [reply src/uct/api/uct.h]: yeah, i though about it as well, but currently all our params (in both UCP and UCT) are "in" only, i think it's better to keep this convention

Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: Agreed. Then should we separate address pointed as hint and the returned address? So change the prototype to:
```
ucs_status_t uct_md_mem_alloc(uct_md_h md, uct_mem_alloc_param_t *param, void **address_p, uct_mem_h *memh_p);
```
and for uct_mem_alloc_params_t should the description change from:
```
    /**
     * If @a address_p content is NULL, the underlying allocation routine will
     * choose the address at which to create the mapping. If @a address_p
     * content is non-NULL but UCT_MD_MEM_FLAG_FIXED is not set, the address
     * will be interpreted as a hint as to where to establish the mapping. If
     * @a address_p is non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then the
     * specified address is interpreted as a requirement. In this case, if the
     * mapping to the exact address cannot be made, the allocation request
     * fails.
     */
    void               **address_p;
```
to
```
    /**
     * If @a address content is NULL, the underlying allocation routine will
     * choose the address at which to create the mapping. If @a address
     * content is non-NULL but UCT_MD_MEM_FLAG_FIXED is not set, the address
     * will be interpreted as a hint as to where to establish the mapping. If
     * @a address is non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then the
     * specified address is interpreted as a requirement. In this case, if the
     * mapping to the exact address cannot be made, the allocation request
     * fails.
     */
    void               *address;
```

Comment (yosefe) [reply src/uct/api/uct.h]: seems excessive to me.

Comment (Akshay-Venkatesh) [reply src/uct/api/uct.h]: I'll leave it as is for now in that case.

File: src/uct/base/uct_mem.c (hunk 14 above)
Comment (Akshay-Venkatesh) [reply src/uct/base/uct_mem.c]: @yosefe I've tried to make sure the param field bits are set before field is used but one assumption I've made is that if PARAM_NAME bit isn't set, then I've used param->name = NULL where applicable. I didn't see checks to make sure that alloc_name is non-NULL, so want to check if this assumption is safe.

Comment (yosefe) [reply src/uct/base/uct_mem.c]: this is not correct, if field_mask is not specifying some field, it could even not be allocated or accessible in memory

File: test/gtest/uct/test_md.cc [hunk 19]
```diff
@@ -130,7 +130,7 @@ UCS_TEST_SKIP_COND_P(test_md, rkey_ptr,
     uct_mem_h memh;
     uct_rkey_bundle_t rkey_bundle;
     unsigned i;
-    uct_mem_alloc_param_t param;
+    uct_mem_alloc_params_t param;
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:133]: param->params (in all cases below)

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> need to fix the params access w/o checking field_mask in all places

@yosefe by this you mean that we need to check *up front* at the start of the function and not every instance where the fields are accessed, right? If so, then does it make sense to assign variables with values of params to user passed or default (NULL/0/whatever relevant) values at the start of the function; and then use variables everywhere else in the function. This would minimize checks to one check per field. Does that sound reasonable or did you have something else in mind?

**Comment by yosefe** [comment]
@Akshay-Venkatesh we need to check for each parameter before it's accesses
see ucp_mem_map() implementation for example

**Comment by Akshay-Venkatesh** [comment]
Actually that's what I was thinking of as well. Something similar to `ucp_mem_map_check_and_adjust_params` right? I was thinking along the lines of this:
```diff
diff --git a/src/uct/base/uct_mem.c b/src/uct/base/uct_mem.c
index 69676d07a..20a654b43 100644
--- a/src/uct/base/uct_mem.c
+++ b/src/uct/base/uct_mem.c
@@ -56,9 +56,27 @@ static inline int uct_mem_get_mmap_flags(unsigned uct_mmap_flags)
     return mm_flags;
 }
+#define fill_param(_user_param, _filled_param, _field_name, _mask, _default) \
+    _filled_param->_field_name = (_user_param->field_mask &                  \
+                                  UCT_MEM_ALLOC_PARAM_FIELD__mask)           \
+                                 ? _user_param->_field_name : _default;
+
+static void uct_mem_fill_params(const uct_mem_alloc_params_t *param,
+                                 uct_mem_alloc_params_t *fill_param)
+{
+    fill_param(param, fill_param, flags, FLAGS, 0);
+    fill_param(param, fill_param, address_p, ADDR_PTR, NULL);
+    fill_param(param, fill_param, length_p, LENGTH_PTR, NULL);
+    fill_param(param, fill_param, mem_type, MEM_TYPE, UCS_MEMORY_TYPE_HOST);
+    fill_param(param, fill_param, methods, METHODS, NULL);
+    fill_param(param, fill_param, mds, MDS, NULL);
+    fill_param(param, fill_param, name, NAME, NULL);
+}
+
 ucs_status_t uct_mem_alloc(const uct_mem_alloc_params_t *param,
                            uct_allocated_memory_t *mem)
 {
+    uct_mem_alloc_params_t filled_params;
 #ifdef ENABLE_MEMTRACK
     const char *alloc_name = (param->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_NAME)
                              ? param->name : NULL;
@@ -79,6 +97,8 @@ ucs_status_t uct_mem_alloc(const uct_mem_alloc_params_t *param,
     ssize_t huge_page_size;
 #endif
+    uct_mem_fill_params(param, &filled_params);
+
     if ((param->field_mask & UCT_MEM_ALLOC_PARAM_FIELD_LENGTH_PTR) &&
         (*param->length_p == 0)) {
         goto err_invalid_length;
```

Later in the function, I can check values of `filled_params` and handle different cases. I can modify the above to what ucp_mem_map does. Does that sound alright?

**Comment by Akshay-Venkatesh** [comment]
I see some messages like this:
```
[1595973321.807183] [swx-rdmz-ucx-althca:13131:0]         parser.c:1629 UCX  WARN  unused env variables: UCX_UD_TIMEOUT, UCX_RC_TIMEOUT,... (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
[0x7228e740] local address length: 255
Waiting for connection...
[0x7228e740] receive handler called with status 0 (Success), length 263
finish to receive UCX address message
[1595973341.836972] [swx-rdmz-ucx-althca:13131:0]         tcp_ep.c:765  UCX  ERROR tcp_ep 0x7f10680008c0 (state=CONNECTED): send(14) failed: Input/output error
unable to send UCX data message (Input/output error)
[0x7228e740] failure handler called with status -80 (Endpoint timeout)
```
But other than that there are segfaults at end of test logs that seems to be failing at different points.

@yosefe Can we retest this?

**Comment by Akshay-Venkatesh** [comment]
@yosefe Thoughts on why tests are failing randomly?

**Comment by yosefe** [comment]
@Akshay-Venkatesh these failures are probably due to the changes in this PR. have you tried to run the same tests on a local machine?

**Comment by Akshay-Venkatesh** [comment]
@yosefe I did check these tests and didn't see these failures. I'll double check on a different platform.

**Comment by Akshay-Venkatesh** [comment]
The individual tests that are failing here don't seem to be failing on a different platform. I see that the tests that have failed now are different from the tests that had failed last time. I'll report back with results of running the full set of tests from gtest and see if there's an issue elsewhere.

**Comment by yosefe** [comment]
replaced by #5562
